lib/host/wasi_crypto/symmetric/eae/hkdf.cpp

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "host/wasi_crypto/symmetric/eae/hkdf.h"
+#include "host/wasi_crypto/symmetric/alg.h"
+#include "host/wasi_crypto/symmetric/factory.h"
+#include "host/wasi_crypto/utils/error.h"
+#include "host/wasi_crypto/utils/evp_wrapper.h"
+#include "host/wasi_crypto/utils/secret_key.h"
+#include "openssl/kdf.h"
+#include "openssl/rand.h"
+#include "wasi_crypto/api.hpp"
+
+namespace WasmEdge {
+namespace Host {
+namespace WasiCrypto {
+namespace Symmetric {
+
+template <int ShaNid> constexpr uint32_t Hkdf<ShaNid>::getKeySize() noexcept {
+  if constexpr (ShaNid == NID_sha256)
+    return 32;
+  if constexpr (ShaNid == NID_sha512)
+    return 64;
+}
+
+template <int ShaNid> constexpr void *Hkdf<ShaNid>::getShaCtx() noexcept {
+  return static_cast<void *>(const_cast<EVP_MD *>(EVP_get_digestbynid(ShaNid)));
+}
+
+template <int ShaNid>
+constexpr Algorithm Hkdf<ShaNid>::getExpandAlg() noexcept {
+  if constexpr (ShaNid == NID_sha256)
+    return Algorithm::HkdfSha256Expand;
+  if constexpr (ShaNid == NID_sha512)
+    return Algorithm::HkdfSha512Expand;
+}
+
+template <int ShaNid>
+WasiCryptoExpect<typename Hkdf<ShaNid>::Expand::Key>
+Hkdf<ShaNid>::Expand::Key::generate(OptionalRef<Options>) noexcept {
+  return WasiCryptoUnexpect(__WASI_CRYPTO_ERRNO_UNSUPPORTED_FEATURE);
+}
+
+template <int ShaNid>
+WasiCryptoExpect<typename Hkdf<ShaNid>::Expand::Key>
+Hkdf<ShaNid>::Expand::Key::import(Span<const uint8_t> Raw) noexcept {
+  ensureOrReturn(Raw.size() == getKeySize(), __WASI_CRYPTO_ERRNO_INVALID_KEY);
+  return SecretKey::random(getKeySize());
+}
+
+template <int ShaNid>
+WasiCryptoExpect<typename Hkdf<ShaNid>::Expand::State>
+Hkdf<ShaNid>::Expand::State::open(Key &Key, OptionalRef<Options>) noexcept {
+  return openStateImpl(Key.ref(), EVP_PKEY_HKDEF_MODE_EXPAND_ONLY);
+}
+
+template <int ShaNid>
+WasiCryptoExpect<void>
+Hkdf<ShaNid>::Expand::State::absorb(Span<const uint8_t> Data) noexcept {
+  opensslAssuming(
+      EVP_PKEY_CTX_add1_hkdf_info(Ctx.get(), Data.data(), Data.size()));
+  return {};
+}
+
+template <int ShaNid>
+WasiCryptoExpect<void>
+Hkdf<ShaNid>::Expand::State::squeeze(Span<uint8_t> Out) noexcept {
+  ensureOrReturn(
+      EVP_PKEY_derive(Ctx.get(), Out.data(), addressOfTemporary(Out.size())),
```

Comment:
Why do not create a temporary variable instead of using this dangerous function `addressOfTemporary` (This might be a undefined behaviour)?

Reply:
In previous implementation there are many place need `addressOfTemporary` (e.g `i2d` without bio), reduce temp variable that need naming. After refactoring here is the only location, I will remove it.